server/modlog.ts

```diff
@@ -36,284 +32,279 @@ const PUNISHMENTS = [
+				query += ` AND regex(?, note)`;
+				args.push(this.generateRegex(search.note.searches[0], search.note.isExact));
+			} else {
+				const parts = [];
+				for (const noteSearch of search.note.searches) {
+					parts.push(`regex(?, note)`);
+					args.push(this.generateRegex(noteSearch, search.note.isExact));
+				}
+				query += ` AND (${parts.join(' OR ')})`;
+			}
 		}
-		for (const fileName of output.stdout.split('\n').reverse()) {
-			if (fileName) results.insert(fileName);
+
+		query += ` ORDER BY timestamp DESC`;
+		if (maxLines) {
+			query += ` LIMIT ?`;
+			args.push(maxLines);
 		}
-		return results;
+		return {statement: this.database.prepare(query), args};
 	}
 
 	async getGlobalPunishments(user: User | string, days = 30) {
-		const response = await PM.query({
-			rooms: ['global' as ModlogID],
-			search: toID(user),
-			isExact: true,
-			maxLines: days * 10,
-			onlyPunishments: 'global',
-		});
-		return response.length;
+		const userid = toID(user);
+		const args: (string | number)[] = [
+			userid, userid, userid, (Date.now() / 1000) - (days * 24 * 60 * 60), ...GLOBAL_PUNISHMENTS,
+		];
+		const results = await this.runSQLWithResults({statement: this.globalPunishmentsSearchQuery, args});
+		return results.length;
 	}
 
 	async search(
-		roomid: ModlogID = 'global', search = '', maxLines = 20, exactSearch = false, onlyPunishments = false
+		roomid: ModlogID = 'global',
+		search: ModlogSearch = {},
+		maxLines = 20,
+		onlyPunishments = false
 	): Promise<ModlogResults> {
-		const rooms = (roomid === 'public' ?
-			[...Rooms.rooms.values()]
-				.filter(room => !room.settings.isPrivate && !room.settings.isPersonal)
-				.map(room => room.roomid) :
-			[roomid]);
-
-		const query = {
-			rooms: rooms,
-			search: search,
-			isExact: exactSearch,
-			maxLines: maxLines,
-			onlyPunishments: onlyPunishments,
-		};
-		const response = await PM.query(query);
-
-		if (response.duration > LONG_QUERY_DURATION) {
-			Monitor.log(`Long modlog query took ${response.duration} ms to complete: ${query}`);
-		}
-		return {results: response, duration: response.duration};
-	}
-
-	private async readRoomModlog(path: string, results: SortedLimitedLengthList, regex?: RegExp) {
-		const fileStream = FS(path).createReadStream();
-		let line;
-		while ((line = await fileStream.readLine()) !== null) {
-			if (!regex || regex.test(line)) {
-				results.insert(line);
-			}
+		const rooms = (roomid === 'public' || roomid === 'all' ?
+			[...Rooms.rooms.values(), {roomid: 'global', settings: {isPrivate: false, isPersonal: false}}]
+				.filter(room => {
+					if (roomid === 'all') return true;
+					return !room.settings.isPrivate && !room.settings.isPersonal;
+				})
+				.map(room => room.roomid as ModlogID) :
+			[roomid]
+		);
+
+		const query = this.prepareSearch(rooms, maxLines, onlyPunishments, search);
+		const start = Date.now();
+		const rows = await this.runSQLWithResults(query);
+		const results: ModlogEntry[] = rows.map((row: any) => {
```

Comment:
chained `map` + `filter` is inefficient if you plan to return a non-trivial # of results